watertap/unit_models/tests/test_unit_test_harness.py

```diff
@@ -0,0 +1,111 @@
+)
+import watertap.property_models.NaCl_prop_pack as props
+from watertap.unit_models.tests.unit_test_harness import UnitTestHarness
+
+from idaes.core.solvers import get_solver
+from idaes.core.util.scaling import calculate_scaling_factors
+
+# -----------------------------------------------------------------------------
+# Get default solver for testing
+solver = get_solver(options={"bound_push": 1e-8})
+
+
+# -----------------------------------------------------------------------------
+class TestReverseOsmosis0D_default(UnitTestHarness):
+    def configure(self):
+        # self.prop_pack = props.NaClParameterBlock
+        # self.param_args = {}
+        # build unit
+        m = ConcreteModel()
+        m.fs = FlowsheetBlock(dynamic=False)
+        m.fs.properties = props.NaClParameterBlock()
+        m.fs.unit = ReverseOsmosis0D(
+            property_package=m.fs.properties,
+            has_pressure_change=True,
+            concentration_polarization_type=ConcentrationPolarizationType.calculated,
+            mass_transfer_coefficient=MassTransferCoefficient.calculated,
+        )
+
+        # fully specify system
+        feed_flow_mass = 1
+        feed_mass_frac_NaCl = 0.035
+        feed_mass_frac_H2O = 1 - feed_mass_frac_NaCl
+        feed_pressure = 50e5
+        feed_temperature = 273.15 + 25
+        membrane_pressure_drop = 3e5
+        length = 20
+        membrane_area = 50
+        A = 4.2e-12
+        B = 3.5e-8
+        pressure_atmospheric = 101325
+
+        m.fs.unit.inlet.flow_mass_phase_comp[0, "Liq", "NaCl"].fix(
+            feed_flow_mass * feed_mass_frac_NaCl
+        )
+        m.fs.unit.inlet.flow_mass_phase_comp[0, "Liq", "H2O"].fix(
+            feed_flow_mass * feed_mass_frac_H2O
+        )
+        m.fs.unit.inlet.pressure[0].fix(feed_pressure)
+        m.fs.unit.inlet.temperature[0].fix(feed_temperature)
+        m.fs.unit.deltaP.fix(-membrane_pressure_drop)
+        m.fs.unit.area.fix(membrane_area)
+        m.fs.unit.A_comp.fix(A)
+        m.fs.unit.B_comp.fix(B)
+        m.fs.unit.permeate.pressure[0].fix(pressure_atmospheric)
+
+        m.fs.unit.feed_side.channel_height.fix(0.002)
+        m.fs.unit.feed_side.spacer_porosity.fix(0.75)
+        m.fs.unit.length.fix(length)
+
+        # scale unit
+        m.fs.properties.set_default_scaling(
+            "flow_mass_phase_comp", 1, index=("Liq", "H2O")
+        )
+        m.fs.properties.set_default_scaling(
+            "flow_mass_phase_comp", 1e2, index=("Liq", "NaCl")
+        )
+        calculate_scaling_factors(m)
+
+        self.unit_model_block = m.fs.unit
+        self.unit_statistics = {
+            "number_variables": 121,
+            "number_total_constraints": 109,
+            "number_unused_variables": 0,
+        }
+        self.unit_solution = {
+            (
+                "feed_side",
+                "properties_in",
+                "flow_mass_phase_comp",
+                ("Liq", "H2O"),
+            ): 0.965,
+            (
```

Comment:
@TimBartholomew I believe we wanted it to look like this, but I could never get it to work.
`        self.unit_solution = {
            (m.fs.feed_side_properties_in.flow_mass_phase_comp["Liq", "H2O"]): 0.965,`

My current workaround is functional, but far from ideal imo...

Reply:
@MarcusHolly do you remember what was the error you were getting when trying to use the Pyomo object as a dict key directly? I've tried with `m.fs.unit.inlet.flow_mass_phase_comp[0, "Liq", "H2O"]` and I'm getting `TypeError: unhashable type: '_GeneralVarData'`.

I've quickly tried to check if there's a way to "drill down" a Pyomo object from a string (sort of like a "deep `getattr`) e.g. in this case

```py
obj_1 = m.fs.unit.inlet.flow_mass_phase_comp[0, "Liq", "H2O"]
obj_2 = m["fs.unit.inlet.flow_mass_phase_comp[0, 'Liq', 'H2O']"]
assert obj_1 is obj_2
```

I wasn't able to find much but it's entirely possible that this is due to my very superficial knowledge of Pyomo, so I'll defer to @bknueven for a proper answer.

If this functionality is not natively available, we should still be able to find an alternative so reduce the complexity of the current solution based on `if len(key) ...`. For example:

1. In `self.configure()`, define `self.unit_solution` using string keys:
  ```py
  self.unit_solutions = {
      m.fs.unit.inlet.flow_mass_phase_comp[0, "Liq", "H2O"]: 0.965,
      ...
  }
  ```
2. When testing, use `eval()` to "extract" the Pyomo object back from the `ConcreteModel` instance:
  ```py
  for key, expected_val in self.unit_solutions.items():
      pyo_obj = eval(key, {"m": self.model}
      assert pytest.approx(val(pyo_obj), expected_val, ...)
  ```

Reply:
Well part of the complexity is that the unit model test harness is intended to be applied to all unit models, so we want to avoid hard-coding in any `m.fs.unit...` inside of `unit_test_harness.py` because the list would be almost infinitely long. I'm just using this RO example as the first implementation, but, ideally, this should work for all unit models.

Reply:
I see what you mean, and I think this is compatible with what I was proposing. The model_specific keys would be defined only in the derived class, leaving `unit_test_harness.py` completely model-agnostic.

In other words, in the model-specific file `test_my_model.py` you'd have

```py
from watertap.unit_models.tests.unit_test_harness import UnitTestHarness

class TestForMyModel(UnitTestHarness):
    def configure(self):
        ...
        self.unit_solutions = {
             "m.fs.unit.inlet.flow_mass_phase_comp[0, 'Liq', 'H2O']": 0.965,
             # rest of the model-specific keys
         ...
     }
```

In the model-agnostic `watertap/unit_models/tests/unit_test_harness.py`

```py
class UnitTestHarness:

    def test_solution(self):
    ...
    for key, expected_val in self.unit_solutions.items():
        pyo_obj = eval(key, {"m": self.model})
        assert pytest.approx(val(pyo_obj), expected_val, ...)
```

As you can see, the base `UnitTestHarness` class doesn't need to know anything about the specific model being tested (apart from assuming that `self.model` refers to the `ConcreteModel` instance).

Reply:
I've tried my hand at this method, but I definitely missed something because I'm seeing ` AttributeError: 'TestReverseOsmosis0D_default' object has no attribute 'model'`

Reply:
Oh I didn't realize my comment below was already discussed here